socks/socks4.py

```diff
@@ -39,10 +49,17 @@ def dumps(self) -> bytes:
 class SOCKS4Reply(typing.NamedTuple):
     reply_code: SOCKS4ReplyCode
     port: int
-    addr: bytes
+    addr: str
 
-    def loads(self, data: bytes) -> "SOCKS4Reply":
-        raise NotImplementedError()
+    @classmethod
+    def loads(cls, data: bytes) -> "SOCKS4Reply":
+        if len(data) != 8:
+            raise ProtocolError("Malformed reply")
+        return cls(
```

Comment:
We should raise a `ProtocolError` if the first byte isn't `0x04`

Reply:
The [protocol definition](https://www.openssh.com/txt/socks4.protocol) says it should be 0:

> VN is the version of the reply code and should be 0

But yeah, we should check for that. The protocol also says the IP and port bytes should be ignored, so I guess we should not raise based if we cannot parse them.